### DISCUSSION TITLE
Policy which only applies to users that are not logged in

### DISCUSSION THREAD

**Discussion by colinramsay** [supabase/supabase#5547] [Questions] [answered]
I would have thought that the following would restrict a select to only logged-out users:

`(role() = 'anon'::text)`

or:

`(role() <> 'authenticated'::text)`

I guess I'm missing something fundamental here?

**Comment by GaryAustin1** [comment]
From Postgres:
.......
Ordinary comparison operators yield null (signifying "unknown"), not true or false, when either input is null. For example, 7 = NULL yields null, as does 7 <> NULL. When this behavior is not suitable, use the IS [ NOT ] DISTINCT FROM constructs:

a IS DISTINCT FROM b
a IS NOT DISTINCT FROM b

For non-null inputs, IS DISTINCT FROM is the same as the <> operator. However, if both inputs are null it returns false, and if only one input is null it returns true. Similarly, IS NOT DISTINCT FROM is identical to = for non-null inputs, but it returns true when both inputs are null, and false when only one input is null. Thus, these constructs effectively act as though null were a normal data value, rather than "unknown".

To check whether a value is or is not null, use the constructs:

expression IS NULL
expression IS NOT NULL
.......

So you probably want (role() IS NULL)

**Comment by colinramsay** [comment]
I appreciate the help. I thought that `role()` would return `anon` rather than NULL when not logged in? Using `role() is NULL` results in no rows being returned.

**Reply by GaryAustin1** [reply]
Sorry was thinking auth.uid().  I always use auth.role() instead of role() even though the UI hides it. The path maybe defined for the anon user, not sure.

**Comment by colinramsay** [comment, marked as answer]
If anyone stumbles on this question looking for a hint, the cause of my issue was simply that I was not setting my auth details before making a request to the database, therefore there were no role or uid details for the policy to work with.
